src/main/java/io/github/moulberry/notenoughupdates/profileviewer/SacksPage.kt

```diff
@@ -0,0 +1,505 @@
+    private var sacksJson = Constants.SACKS
+    private var tooltipToDisplay = listOf<String>()
+    private var currentProfile: SkyblockProfiles.SkyblockProfile? = null
+
+    private var currentSack = "All"
+
+    private var page = 0
+    private var maxPage = 0
+    private val arrowsHeight = 162
+
+    private val columns = 7
+    private val rows = 4
+    private val pageSize = columns * rows
+
+    private var guiLeft = GuiProfileViewer.getGuiLeft()
+    private var guiTop = GuiProfileViewer.getGuiTop()
+
+    private val sackContents = mutableMapOf<String, SackInfo>()
+    private val sackItems = mutableMapOf<String, SackItem>()
+    private val playerRunes = mutableListOf<String>()
+
+    private val sackPattern = "^RUNE_(?<name>\\w+)_(?<tier>\\d)\$".toPattern()
+
+    override fun drawPage(mouseX: Int, mouseY: Int, partialTicks: Float) {
+        guiLeft = GuiProfileViewer.getGuiLeft()
+        guiTop = GuiProfileViewer.getGuiTop()
+
+        MC.textureManager.bindTexture(pv_sacks)
+        Utils.drawTexturedRect(
+            guiLeft.toFloat(),
+            guiTop.toFloat(),
+            instance.sizeX.toFloat(),
+            instance.sizeY.toFloat(),
+            GL11.GL_NEAREST
+        )
+
+        val newProfile = selectedProfile
+        if (newProfile == null) {
+            Utils.drawStringCentered("§cMissing Profile Data", guiLeft + 250, guiTop + 101, true, 0)
+            return
+        }
+
+        if (sacksJson == null) {
+            Utils.drawStringCentered("§cMissing Repo Data", guiLeft + 250, guiTop + 101, true, 0)
+            return
+        }
+
+        if (newProfile != currentProfile) {
+            getData()
+            currentProfile = selectedProfile
+        }
+
+        val currentSackData = sackContents[currentSack] ?: run {
+            Utils.drawStringCentered("§cApi Info Missing", guiLeft + 250, guiTop + 101, true, 0)
+            return
+        }
+        // after this point everything in the constants json exists and does not need to be checked again except "item"
+
+        val name = if (currentSack == "All") "§2All Sacks" else "§2$currentSack Sack"
+        Utils.renderShadowedString(name, (guiLeft + 78).toFloat(), (guiTop + 74).toFloat(), 105)
+
+        Utils.renderAlignedString(
+            "§6Value",
+            "§f${StringUtils.formatNumber(currentSackData.sackValue.toLong())}",
+            (guiLeft + 27).toFloat(),
+            (guiTop + 91).toFloat(),
+            102
+        )
+
+        Utils.renderAlignedString(
+            "§2Items",
+            "§f${StringUtils.formatNumber(currentSackData.itemCount)}",
+            (guiLeft + 27).toFloat(),
+            (guiTop + 108).toFloat(),
+            102
+        )
+
+        GlStateManager.enableDepth()
+
+        var xIndex = 0
+        var yIndex = 0
+
+        val sackTypes = sacksJson.getAsJsonObject("sacks")
+
+        val startIndex = page * pageSize
+        val endIndex = (page + 1) * pageSize
+        var itemCount = -1
```

Comment:
Can we replace this variable by just iterating over sackTypes.entrySet().withIndex() instead?